### ISSUE TITLE
ASCII color codes are displayed inline for for editors

### ISSUE THREAD

**Issue by lindhe** [rust-lang/rust-analyzer#15443]
Sometimes, the error shown in Vim (and apparently VS Code too) is printed with ASCII color codes showing in the string from rust_analyzer. It makes the error very hard to read. See <https://github.com/neovim/nvim-lspconfig/issues/2760>.

It seems to me like `rust-analyzer` does not properly distinguish between being executed in a terminal environment or not, and decide from that whether to print ASCII color codes or skip them.

**In terminal:**

![Screenshot from 2023-08-11 16-16-58](https://github.com/rust-lang/rust-analyzer/assets/7773090/147a81d3-8073-4b6b-92fe-886cca64534c)

**In editor:**

![Screenshot from 2023-08-11 10-51-47](https://github.com/rust-lang/rust-analyzer/assets/7773090/8dde823d-2ac3-4c10-9557-c6066cdf2756)

## Steps to reproduce

1. Create the following crate:

```shell
cargo new -q foo
cd foo
cargo add -q rocket@0.5.0-rc.3
cat <<EOF > src/main.rs
use rocket::UriDisplayPath;

fn main() {
    println!("Hello, world!");
}

#[derive(UriDisplayPath)]
pub enum Foo {
    Bar,
}
EOF
```

2. Optionally run `cargo check`.
3. Open `src/main.rs` with Vim.

## Other info

**rust-analyzer version**: (eg. output of "rust-analyzer: Show RA Version" command, accessible in VSCode via <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>)

[Mason](https://github.com/williamboman/mason.nvim) reports rust-analyzer 2023-08-07.

**rustc version**: (eg. output of `rustc -V`)

```console
$ rustc -V
rustc 1.71.1 (eb26296b5 2023-08-03)
```

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`)

Not sure, nothing I've set explicitly at least. I run NeoVim with [rust-lang/rust.vim](https://github.com/rust-lang/rust.vim) and[ rust_analyzer for lspconfig](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#rust_analyzer) installed via [Mason](https://github.com/williamboman/mason.nvim).

**Comment by Veykril** [comment]
Your error message for some reason contains a green `[note]` part, which is what you are seeing inline. That should not be there as that is not the normal diagnostic output of the compiler. This is what I am getting here 
![image](https://github.com/rust-lang/rust-analyzer/assets/3757771/3a62d0d9-10f8-494d-87de-1e2729a4453c)
I don't know what could cause this for you but something seems to fiddle with your compiler output?

**Comment by tadeokondrak** [comment]
I think the derive macro is emitting escape codes: https://github.com/SergioBenitez/proc-macro2-diagnostics/blob/master/src/line.rs#L79

**Comment by Veykril** [comment]
Ah good catch, that's indeed gonna be the issue. There isn't really anything we can do here then, r-a just spits out the error messages we get in json verbatim as the expectation is for compiler diagnostics to be just text. This might be something to request as a feature for the proc-macro api, since the compiler itself has a flag that toggles whether the json output has colors in it or not which this should hook into.

**Comment by lindhe** [comment]
Ah, there it is! Thanks so much for identifying the real culprit for me! I'll continue my hunt over there.

**Comment by lindhe** [comment]
Interestingly, it seems like `cargo build` and proc-macro properly toggles color output sometimes!

Here's what it looks like in my terminal, and the color codes are not printed!

![Screenshot from 2023-08-12 10-01-05](https://github.com/rust-lang/rust-analyzer/assets/7773090/9aef8114-9714-4128-955e-ffd6ad45b3dd)

It seems like proc-macro do attempt to use different logic depending on [whether or not `color` is set](https://github.com/SergioBenitez/proc-macro2-diagnostics/blob/45fa2d68669d10daeaf0f9b664a96576626fae16/src/line.rs#L78C1-L78C1):

> ```rust
> #[cfg(all(feature = "colors", not(nightly_diagnostics)))]
> ```

@Veykril, you mentioned "the error messages we get in json". It  seems like [`cargo build --message-format`](https://doc.rust-lang.org/cargo/commands/cargo-build.html#option-cargo-build---message-format) has a few different options that might be of relevance here: `json`, `json-diagnostic-short`, `json-diagnostic-rendered-ansi` and `json-render-diagnostics`.

What field do you render? Is it `message.message`?

BTW, I notice that `cargo build` prints the `[note]` while `rustc` does not. What do `rust-analyzer` use internally to build/check?

**Comment by lindhe** [comment]
Rats! I think this might be a bug in Cargo. Even with `--color=never`, the `color` feature appears to be set when building using `--json`:

![Screenshot from 2023-08-12 11-50-16](https://github.com/rust-lang/rust-analyzer/assets/7773090/867a83a0-0e88-432a-b0ed-36433ddc2590)

**Comment by Veykril** [comment]
No, there is no bug here. The colored `[note]` is created by the proc-macro itself, no tool has control over that proc-macro emitting color codes or not. Hence why I said

> This might be something to request as a feature for the proc-macro api, since the compiler itself has a flag that toggles whether the json output has colors in it or not which this should hook into.

earlier

**Comment by lindhe** [comment]
> the compiler itself has a flag that toggles whether the json output has colors in it or not

What flag is that, if not `--color`?

**Comment by Veykril** [comment]
That flag controls the color of the compilers output colors for diagnostic, the color you are seeing here (that is the `[note]` part) is created by the proc-macro though, not by the compiler. The compiler merely emits the message that the proc-macro tells it to.

**Comment by lindhe** [comment]
Yes, but proc-macro emits that conditionally on whether or not the `color` feature is set:

https://github.com/SergioBenitez/proc-macro2-diagnostics/blob/45fa2d68669d10daeaf0f9b664a96576626fae16/src/line.rs#L78C1-L78C1

**Comment by Veykril** [comment]
Yes, that is a feature flag. Passing `--color` or not to the compiler is rrelevant to that though, as that is not a flag that sets the feature. In fact, crates cannot observe `--color` being passed or not whatsoever

**Comment by lindhe** [comment]
Oh, that's interesting. Any idea how on earth proc-macro is able to distinguish between colorized and plain output then? Because, as you can see in my screenshots, it does not colorize "[note]" when I pipe it to `cat`.

**Comment by lindhe** [comment]
> the compiler itself has a flag that toggles whether the json output has colors in it or not which this should hook into.

Do you have any pointers to how one checks that flag?

**Comment by lindhe** [comment]
> That flag controls the color of the compilers output colors for diagnostic, the color you are seeing here (that is the `[note]` part) is created by the proc-macro though, not by the compiler. The compiler merely emits the message that the proc-macro tells it to.

> Passing `--color` or not to the compiler is rrelevant to that though, as that is not a flag that sets the feature. In fact, crates cannot observe `--color` being passed or not whatsoever

I am sure you are right, I just don't understand. What trips me up here is that `[note]` is printed _in color_ when setting `--color=always` but printed _not in color_ when setting `--color=never` (see screenshot below). Unless Cargo or the compiler does a text replacement to strip color codes from the output, which I very much doubt, I fail to see how it can be the case that `--color=always` is not picked up by the proc-macro.

![Screenshot from 2023-08-12 19-03-46](https://github.com/rust-lang/rust-analyzer/assets/7773090/aa31774e-76c2-4ae8-afc5-8684ff496d47)

**Comment by SergioBenitez** [comment]
All, copying my comments from https://github.com/SergioBenitez/proc-macro2-diagnostics/issues/3#issuecomment-1712314620 here for discourse:

> The issue [...] is that as far as I can tell, there's no way for a proc-macro to know if it should emit colors or not. The proxy is typically to check whether we're connected to a TTY, but Cargo captures all output, so we're never connected to a TTY, and detecting via this manner would mean never emitting colors, which would be a shame.
>
> Ideally, there would be some way to query 1) whether cargo itself will be emitting colors and/or 2) what message-format is expected. Alternatively, if we could inquire about how the compiler is invoked, we might be able to patch this minimally for rust-analyzer. It doesn't seem like 1) or 2) are viable options now, but the alternative might be, given that rust-analyzer appears to set RUSTC_WRAPPER to rust-analyzer by default. If we can detect this, we can stop emitting colors when rust-analyzer invokes rustc.
>
>  [...] the above didn't work. First, the wrapper is only set for build scripts, which is fine-ish as we can forward the env-var along via Cargo. Second, and more importantly, the wrapper is only set the first time the build script is built/run, so the solution only really works once in a given workspace, if it works at all.
>
> Until there's some reliable mechanism to detect whether we should color, I think the "correct" solution here is for rust-analyzer to strip ANSI escapes from any output before it re-emits it. Cargo itself uses https://crates.io/crates/strip-ansi-escapes, which is why using --color=never works as expected.

**Comment by Veykril** [comment]
Interesting, so cargo already handles this specially itself as well? In that case you are right, we should do the same as cargo here when processing `compile_error!` invocations (and also we should pass `--color=never` to cargo check, I don't think we are doing that currently)

**Comment by DriedYellowPeach** [comment]
Is there any solution I can avoid this right now? I tried add color=never in my .toml, but it only avoid color output in cargo build, not in diagnostics. Can I setup rust-analyzer or something?

**Comment by lindhe** [comment]
I know of no work-around  until they fix this.

**Comment by SergioBenitez** [comment]
One "work-around" is to use a nightly toolchain, at least for rust-analyzer.

**Comment by lindhe** [comment]
After some time not running any Rust, I'm back at this and notice the issue is still around. Is the best work-around still to use the nightly toolchain? Would that risk disrupting other tools, or can I run only `rust-analyzer` in nightly and use everything else at stable?

**Comment by ShoyuVanilla** [comment]
I'll try striping ANSI escape codes from diagnostics
